aip/client-libraries/4234.md

```diff
@@ -0,0 +1,103 @@
+---
+aip:
+  id: 4234
+  scope: client-libraries
+  state: draft
+  created: 2019-06-28
+  updated: 2019-07-01
+permalink: /client-libraries/4234
+redirect_from:
+  - /4234
+---
+
+# Bazel protobuf targets
+
+When using Bazel to compile protobuf and generate message libraries and RPC service
+libraries, the naming choice for these targets has implications on derivative targets, and
+those target names have implications for per-language integrations. By applying a common
+approach to how individual `.proto` files are combined into targets, and to what those
+targets are named, clients can have a uniform experience from API to API.
+
+Some of the various language constraints are as follows:
+
+*   Java's `option java_outer_classname`: per the API style guide is the name of
+    the proto file, with "Proto" appended. Unless `option java_multiple_files` is true,
+    the file name defines the packaging of the types.
+*   Go's target names will define the import path for the result. However, Go's package
+    names have a more restrictive character set (e.g., no dashes), and in some cases, that
+    means that the import path and package name will disagree. This interferes with some
+    tooling.
+
+## Guidance
+
+### Targets
+
+In the following file name cases, if the service name or resource name is more than one
+word, they should be separated by underscores.
+
+#### Services and request/response messages
+
+Individual services along with their request and response messages **should** live in
+separate `proto_library` targets, named `{service}_service_proto`.
+
+**NOTE:** Services **should** not share request or response messages. Reuse of resources
+or common messages is preferred.
+    
+#### Resource messages
+
+*   Resources that are a primary resource for an individual service, **should** be 
+    in a `proto_library` target named `{resource}_proto`.
+*   Otherwise, per-service resources **should** be in a `proto_library` target named 
+    `{service}_resources_proto`.
+*   If there are remaining shared resources, they **should** be in a `proto_library`
+    target named `resources_proto`.
+    
+#### Additional messages
+
+Remaining necessary messages **should** be in a `proto_library` target named
+`common_proto`.
+
+### `.proto` file naming convention
+
+In the case where there's a single `.proto` file in the `srcs` of the
```

Comment:
This seems out of scope for this AIP, and it seems to contradict the current style guide. Am I missing something?

Reply:
Do you have a pointer to the style guide where it states otherwise? The point here was to not have unnecessarily disagreement between .proto source files's name and the `proto_library` target name, especially since the source file has an impact on the `java_outer_classname`.

Reply:
Sure, the current style guide advice is here: https://cloud.google.com/apis/design/file_structure

Reply:
Perhaps this should then be restructured to derive from that, or at least describe what bundling of multiple .proto targets is desirable vs. not. Would the style guide be amenable to having additional guidance about what constitutes "moving into separate files as needed"?

Reply:
I wrote an AIP for this that you can use as the source of truth: aip.dev/191